integration/tests/epochs/epoch_test.go

```diff
@@ -142,68 +137,40 @@ func (s *Suite) TestEpochJoin() {
-	require.Contains(s.T(), proposedTable.(cadence.Array).Values, cadence.String(info.NodeID.String()), fmt.Sprintf("expected new node to be in proposed table: %x", info.NodeID))
-
-	containerConfigs := []func(config *testnet.NodeConfig){
-		testnet.WithLogLevel(zerolog.DebugLevel),
-		testnet.WithID(info.NodeID),
-	}
-
-	nodeConfig := testnet.NewNodeConfig(role, containerConfigs...)
-	testContainerConfig := testnet.NewContainerConfig(info.ContainerName, nodeConfig, info.NetworkingKey, info.StakingKey)
-	err := testContainerConfig.WriteKeyFiles(s.net.BootstrapDir, flow.Localnet, info.MachineAccountAddress, encodable.MachineAccountPrivKey{PrivateKey: info.MachineAccountKey}, role)
-	require.NoError(s.T(), err)
-
-	s.WaitForPhase(ctx, flow.EpochPhaseSetup)
-	snapshot, err := s.client.GetLatestProtocolSnapshot(ctx)
-	require.NoError(s.T(), err)
-
-	//add our container to the network
-	err = s.net.AddNode(s.T(), s.net.BootstrapDir, testContainerConfig)
-	require.NoError(s.T(), err, "failed to add container to network")
+	s.net.StopContainers()
+}
 
-	//start our test container
-	testContainer := s.net.ContainerByID(info.NodeID)
-	testContainer.WriteRootSnapshot(snapshot)
-	testContainer.Container.Start(ctx)
+// TestEpochLeave this test add a new node to the network and after some time un-stake and remove a node
+// from the network of the same node type. The test then makes the assertions necessary for the network
+// to be considered healthy
+func (s *Suite) TestEpochLeave() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
 
-	// wait for new container to startup and start processing blocks
-	// wait for end of epoch
-	epochFinalView, err := snapshot.Epochs().Current().FinalView()
-	require.NoError(s.T(), err)
-	s.BlockState.WaitForSealedView(s.T(), epochFinalView)
+	env := utils.LocalnetEnv()
+	role := flow.RoleAccess
 
-	// get head of snapshot we bootstrap from
-	bootstrapHead, err := snapshot.Head()
-	require.NoError(s.T(), err)
+	// grab the first container of this node role type, this is the container we will replace
+	containerToReplace := s.net.ContainersByRole(role)[0]
 
-	// get snapshot directly from new AN and compare head with head from the
-	// snapshot that was used to bootstrap the node
-	clientAddr := fmt.Sprintf(":%s", s.net.AccessPortsByContainerName[info.ContainerName])
-	client, err := testnet.NewClient(clientAddr, s.net.Root().Header.ChainID.Chain())
-	require.NoError(s.T(), err)
-	snapshot, err = client.GetLatestProtocolSnapshot(ctx)
-	require.NoError(s.T(), err)
+	// stake a new node and start the docker container
+	info, snapshot, _ := s.StakeAndStartNewNode(ctx, env, role)
 
-	head, err := snapshot.Head()
+	// at this point our new node has started and is participating
+	// in the protocol, now we can remove the old node from the protocol and pause the container
+	result, err := s.SubmitAdminRemoveNodeTx(ctx, env, containerToReplace.Config.NodeID)
```

Comment:
```
       A                             B
|--------------|-------|---------||---------
 Staking.       Setup.  Commited.   Staking
```

Currently we are removing the leaving node at point B (staking phase of epoch 2) when we should be removing the leaving node at point A (staking phase of epoch 1) along with staking the joining node, so that both changes take effect at the epoch 1->2 transition

Reply:
https://github.com/onflow/flow-go/pull/1728/commits/657e37d5bb69949bc8b30a6ca1acf70570779c71